From 76248b2b387497c105ec9e3a20ac6ce5aa538771 Mon Sep 17 00:00:00 2001
From: nico <nicorg2529@gmail.com>
Date: Tue, 2 Oct 2018 06:43:34 +0000
Subject: [PATCH 4/7] CameraService: Initialize CameraParameters for the
 cameras and cache them onFirstRef

To avoid race with other clients and evicting conflicting client when first open camera.

Test: It prevents error "CameraService::connect evicting conflicting client for camera ID %s" when camera is opened for the first time after boot.
Change-Id: I15e33ddbb64a0f1b98d22de144288f9cfe72e7fd
---
 services/camera/libcameraservice/CameraService.cpp | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/services/camera/libcameraservice/CameraService.cpp b/services/camera/libcameraservice/CameraService.cpp
index fb740053e..0716af4d8 100644
--- a/services/camera/libcameraservice/CameraService.cpp
+++ b/services/camera/libcameraservice/CameraService.cpp
@@ -135,6 +135,15 @@ void CameraService::onFirstRef()
     res = enumerateProviders();
     if (res == OK) {
         mInitialized = true;
+
+        std::vector<std::string> deviceIds;
+        deviceIds = mCameraProviderManager->getCameraDeviceIds();
+
+        for (auto& cameraId : deviceIds) {
+            int idInt = atoi(cameraId.c_str());
+            CameraParameters shimParams;
+            getLegacyParametersLazy(idInt, /*out*/&shimParams);
+        }
     }
 
     CameraService::pingCameraServiceProxy();
-- 
2.17.1

